components/style/values/specified/image.rs

```diff
@@ -0,0 +1,370 @@
+                    AngleOrCorner::Angle(Angle(PI * 1.5))
+                },
+                (Some(horizontal), Some(vertical)) => {
+                    AngleOrCorner::Corner(horizontal, vertical)
+                }
+                (None, None) => unreachable!(),
+            }
+        } else if let Ok(angle) = input.try(Angle::parse) {
+            try!(input.expect_comma());
+            AngleOrCorner::Angle(angle)
+        } else {
+            AngleOrCorner::Angle(Angle(PI))
+        };
+
+        Ok(GradientKind::Linear(angle_or_corner))
+    }
+
+    /// Parses a radial gradient from the given arguments.
+    pub fn parse_radial(input: &mut Parser) -> Result<GradientKind, ()> {
+        let shape = input.try(EndingShape::parse)
+                         .unwrap_or(EndingShape::Circle(Size::Keyword(SizeKeyword::FarthestCorner)));
+
+        let position = if input.try(|input| input.expect_ident_matching("at")).is_ok() {
+            try!(Position::parse(input))
+        } else {
+            // TODO(canaltinova): Is it default value? I'm not sure.
+            Position {
+                horiz_keyword: Some(Keyword::Center),
+                horiz_position: None,
+                vert_keyword: Some(Keyword::Center),
+                vert_position: None,
+            }
+        };
+
+        Ok(GradientKind::Radial(shape, position))
+    }
+}
+
+/// Specified values for an angle or a corner in a linear gradient.
+#[derive(Clone, PartialEq, Copy, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum AngleOrCorner {
+    Angle(Angle),
+    Corner(HorizontalDirection, VerticalDirection),
+}
+
+impl ToCss for AngleOrCorner {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        match *self {
+            AngleOrCorner::Angle(angle) => angle.to_css(dest),
+            AngleOrCorner::Corner(horizontal, vertical) => {
+                try!(dest.write_str("to "));
+                try!(horizontal.to_css(dest));
+                try!(dest.write_str(" "));
+                try!(vertical.to_css(dest));
+                Ok(())
+            }
+        }
+    }
+}
+
+/// Specified values for one color stop in a linear gradient.
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct ColorStop {
+    /// The color of this stop.
+    pub color: CSSColor,
+
+    /// The position of this stop. If not specified, this stop is placed halfway between the
+    /// point that precedes it and the point that follows it.
+    pub position: Option<LengthOrPercentage>,
+}
+
+impl ToCss for ColorStop {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        try!(self.color.to_css(dest));
+        if let Some(position) = self.position {
+            try!(dest.write_str(" "));
+            try!(position.to_css(dest));
+        }
+        Ok(())
+    }
+}
+
+define_css_keyword_enum!(HorizontalDirection: "left" => Left, "right" => Right);
```

Comment:
This line was already written in mod.rs actually, but this macro makes things a lot easier :) I used it in `SizeKeyword`